### PR TITLE
Show actual fuel consumption, not a guess

### PR DESCRIPTION
**Pull request by Salty-Panda**
#### Summary
Interface "Show actual fuel consumption, not a guess"

#### Purpose of change

Part of my #46632.
Also, according to @mlangsdorf this function required rewrite anyway.

#### Describe the solution

Make a struct that stores fuel consumption from the last 60 calls of consume_fuel().
Calculate and display average fuel consumption. 
Until now the displayed fuel consumption was calculated based on load, speed, acceleration without regard to actual fuel consumption.

#### Describe alternatives you've considered

I used a solution suggested by Mark.

#### Testing

Drove a few vehicles. Put on a few alternators, crashed in stuff.

### CODE REVIEW

**Comment by actual-nh** [comment]
General build matrix error:
```
src/vehicle.cpp: In member function ‘void vehicle::consume_fuel(int, int, bool)’:
Error: src/vehicle.cpp:4621:22: error: operation on ‘fcd.vehicle::fuel_consumption_data::tail’ may be undefined [-Werror=sequence-point]
Makefile:934: recipe for target 'obj/vehicle.o' failed
 4621 |             fcd.tail = ++fcd.tail % 60;
      |             ~~~~~~~~~^~~~~~~~~~~~~~~~~
Error: src/vehicle.cpp:4623:18: error: operation on ‘fcd.vehicle::fuel_consumption_data::head’ may be undefined [-Werror=sequence-point]
 4623 |         fcd.head = ++fcd.head % 60;
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~
```

**Comment by Salty-Panda** [comment]
Weird, I was told that this would work
https://discord.com/channels/598523535169945603/598529174302490644/808489041221779518

File: src/vehicle.cpp [hunk 1]
```diff
@@ -4623,10 +4606,29 @@ void vehicle::consume_fuel( int load, const int t_seconds, bool skip_electric )
...
-        inserted.first->second += amnt_precise_j;
+
+        //FIFO queue stores fuel consumption in order to calculate average over a minute
+        if( fcd.tail == -1 ) {
+            if( fcd.head == 58 ) {
+                fcd.tail = 0;
+            }
+        } else {
+            fcd.total_fuel -= fcd.fuel_per_sec[ fcd.tail ];
+            fcd.tail = ++fcd.tail % 60;
```
Comment (BrettDong) [inline src/vehicle.cpp:4626]:
```suggestion
            fcd.tail = ( fcd.tail + 1 ) % 60;
```

File: src/vehicle.cpp [hunk 2]
```diff
@@ -4623,10 +4606,29 @@ void vehicle::consume_fuel( int load, const int t_seconds, bool skip_electric )
...
+        }
+        fcd.head = ++fcd.head % 60;
```
Comment (BrettDong) [inline src/vehicle.cpp:4628]:
```suggestion
        fcd.head = ( fcd.head + 1 ) % 60;
```

**Comment by Aivean** [comment]
> Weird, I was told that this would work
> https://discord.com/channels/598523535169945603/598529174302490644/808489041221779518

[It will](http://www.cplusplus.com/reference/map/map/operator[]/).
Check that your struct in initialized correctly when default constructor is used.

**Comment by actual-nh** [comment]
It is also rather clearer to rewrite it the way BrettDong did!

**Comment by actual-nh** [comment]
> UPD: It turns out that error is unrelated to the constructors and initialization. Using `++` and `%` together [could be an undefined behavior](https://stackoverflow.com/a/10623165/1349366).

That being what it's meaning by `operator may be undefined`?

**Comment by Aivean** [comment]
> > UPD: It turns out that error is unrelated to the constructors and initialization. Using `++` and `%` together [could be an undefined behavior](https://stackoverflow.com/a/10623165/1349366).
> 
> That being what it's meaning by `operator may be undefined`?

Yes, the error:
```
 operation on ‘fcd.vehicle::fuel_consumption_data::tail’ may be undefined [-Werror=sequence-point]
Makefile:934: recipe for target 'obj/vehicle.o' failed
```
Is caused by using this construction:   `fcd.head = ++fcd.head % 60;`.  The fix is proposed by @BrettDong above.


---

Another issue is:

```cpp
        auto fuel_data = fuel_used.find( ft );
        if( fuel_data == fuel_used.end() ) {
            fuel_used.insert( { ft, fuel_consumption_data() } );
            fuel_data = fuel_used.find( ft );
        }
        fuel_consumption_data &fcd = fuel_data->second;
```

could be rewritten as:
```cpp
        fuel_consumption_data &fcd = fuel_data[ ft ];
```
Which is what I told on discord and it is unrelated to the former issue.

**Comment by Salty-Panda** [comment]
Thanks, I also reverted to Aivean's code.

File: src/vehicle.h [hunk 3]
```diff
@@ -716,13 +716,21 @@ class vehicle
 
         units::volume total_folded_volume() const;
 
+        struct fuel_consumption_data {
```
Comment (kevingranade) [inline src/vehicle.h:719]: Why isn't this just a std::list<int>?
We need a VERY good reason to step away from standard data structures and start building our own, and I don't see a case for that here.

**Review by kevingranade — CHANGES_REQUESTED** [review]

File: src/vehicle.h (hunk 3 above)
Comment (Salty-Panda) [reply src/vehicle.h:719]: No idea. Mark suggested this precise solution.
I'm not c++ programmer so I assumed some esoteric memory/performance issues.

Although later he mentioned in regard to this PR that his knowledge of libraries is out of date.

If there is a reason to change it to a list is there a specific implementation I should use?|
Or just something like 
```cpp
std::list.push( new_val );
max += new_val;
max -= std::list.first();
std::list.pop();
```
?

Comment (Salty-Panda) [reply src/vehicle.h:719]: I switched to list

File: src/vehicle.cpp [hunk 4]
```diff
@@ -3457,36 +3457,19 @@ int vehicle::basic_consumption( const itype_id &ftype ) const
...
-                debugmsg( "Vehicle unexpectedly has zero acceleration" );
-            } else {
-                amount_pct += 600 * vslowdown / accel;
-            }
-        }
-    }
+
     int energy_j_per_mL = fuel.fuel_energy() * 1000;
-    return -amount_pct * fuel_rate_w / energy_j_per_mL;
+    //If worked for less than 60s, head + 1 is a total amount of stored values
```
Comment (kevingranade) [inline src/vehicle.cpp:3469]:
```suggestion
```

**Comment by Salty-Panda** [comment]
@mlangsdorf 
Not sure when if you want to merge it before release, but as far I can tell it's working fine

**Comment by Salty-Panda** [comment]
Rebased on master, cleared conflicts

**Comment by Salty-Panda** [comment]
Clang tidy error is in shooting test I think? could someone rerun it?
The changes requested by Kevin has been done already, it's just he marked a line without logic that remained unchanged.

**Comment by BrettDong** [comment]
Sadly there isn't a re-run button for this workflow run, perhaps because it was too long ago. You may force-push to this pull request to trigger a rebuild.

File: src/vehicle.cpp [hunk 5]
```diff
@@ -3462,36 +3462,17 @@ int vehicle::basic_consumption( const itype_id &ftype ) const
...
-            } else {
-                amount_pct += 600 * vslowdown / accel;
-            }
-        }
-    }
-    int energy_j_per_mL = fuel.fuel_energy() * 1000;
-    return -amount_pct * fuel_rate_w / energy_j_per_mL;
+
+    int average = fcd.total_fuel / fcd.fuel_per_sec.size();
+    //Fuel energy is in 'kJ', consumption is per 's' so multiply by 3600, then divide by 1000
```
Comment (Salty-Panda) [inline src/vehicle.cpp:3474]:
```suggestion
    //Fuel energy is in 'kJ', consumption is per 's' so multiply by 3600, then divide by 1000 in order to convert kj to ml
```

**Comment by Salty-Panda** [comment]
Looks like something has changed since last test run and small fix is needed, will try to do soon

**Comment by Night-Pryanik** [comment]
Will this fix #37943?

**Comment by Salty-Panda** [comment]
It should remove most inconsistencies in calculating remaining fuel
although it takes average from last minute so might not be too reliable
when repeatedly changing speed or ramming.

**Comment by I-am-Erk** [comment]
I am marking this project as "abandoned" but, since it is doing something we want, I have put it in an Abandoned PR list to hopefully be reinitiated at some point.

**Pull request closed without merging**